test/e2e/framework/metrics/metrics_proxy.go

```diff
@@ -34,37 +34,47 @@ import (
 )
 
 type componentInfo struct {
+	Name string
 	Port int
 	IP   string
 }
 
 // SetupMetricsProxy creates a nginx Pod to expose metrics from the secure port of kube-scheduler and kube-controller-manager in tests.
 func SetupMetricsProxy(c clientset.Interface) error {
-	podList, err := c.CoreV1().Pods(metav1.NamespaceSystem).List(context.TODO(), metav1.ListOptions{})
-	if err != nil {
-		return err
-	}
 	var infos []componentInfo
-	for _, pod := range podList.Items {
-		switch {
-		case strings.HasPrefix(pod.Name, "kube-scheduler-"):
-			infos = append(infos, componentInfo{
-				Port: kubeSchedulerPort,
-				IP:   pod.Status.PodIP,
-			})
-		case strings.HasPrefix(pod.Name, "kube-controller-manager-"):
-			infos = append(infos, componentInfo{
-				Port: kubeControllerManagerPort,
-				IP:   pod.Status.PodIP,
-			})
+	// The component pods might take some time to show up.
+	err := wait.PollImmediate(time.Second*5, time.Minute*5, func() (bool, error) {
+		podList, err := c.CoreV1().Pods(metav1.NamespaceSystem).List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return false, fmt.Errorf("list pods in ns %s: %w", metav1.NamespaceSystem, err)
 		}
-		if len(infos) == 2 {
-			break
+		var foundComponents []componentInfo
+		for _, pod := range podList.Items {
+			switch {
+			case strings.HasPrefix(pod.Name, "kube-scheduler-"):
+				foundComponents = append(foundComponents, componentInfo{
+					Name: pod.Name,
+					Port: kubeSchedulerPort,
+					IP:   pod.Status.PodIP,
+				})
+			case strings.HasPrefix(pod.Name, "kube-controller-manager-"):
+				foundComponents = append(foundComponents, componentInfo{
+					Name: pod.Name,
+					Port: kubeControllerManagerPort,
+					IP:   pod.Status.PodIP,
+				})
+			}
 		}
-	}
-	if len(infos) == 0 {
-		klog.Warningf("Can't find any pods in namespace %s to grab metrics from", metav1.NamespaceSystem)
-		return nil
+		if len(foundComponents) != 2 {
+			klog.Infof("Only %d components found. Will retry.", len(foundComponents))
+			klog.Infof("Found components: %v", foundComponents)
+			return false, nil
+		}
+		infos = foundComponents
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("missing component pods: %w", err)
 	}
 
 	const name = metricsProxyPod
```

Comment:
this pod has a well known name, can we run it only when it is necessary?
```
if metricsProxyPod {
    return nil
}
```
I'm afraid that we can start a precedent with this approach and new components and more  "helpers" pods at the beginning of the e2e,despite most of them doesn't need it
i.e. a Conformance run will create this pod, despite it really doesn't need it IIUIC

Reply:
This is pod is only required if the tests need to fetch metrics from the scheduler or controller-manager, but it is unclear to me how to know if the tests need to grab metrics here.

Reply:
I was assuming this was only using the grabber so, I maybe wrongly, assumed that we can do it as part of the initialisation of the `MetricsGrabber`


```go
// NewMetricsGrabber returns new metrics which are initialized.
func NewMetricsGrabber(c clientset.Interface, ec clientset.Interface, kubelets bool, scheduler bool, controllers bool, apiServer bool, clusterAutoscaler bool) (*Grabber, error) {

   SetupMetricsProxy(c)

}
```
and inside SetupMetricsProxy(), only create the pod if it doesn't exist

Reply:
yeah, I think we could do that.